src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -530,56 +584,105 @@ impl<'tcx> Tree {
+        // transition in more context for diagnostics.
+        let err_handler = |perms_range: Range<u64>,
+                           args: ErrHandlerArgs<'_, TransitionError>|
+         -> InterpError<'tcx> {
+            let ErrHandlerArgs { error_kind, conflicting_info, accessed_info } = args;
+            TbError {
+                conflicting_info,
+                access_cause,
+                error_offset: perms_range.start,
+                error_kind,
+                accessed_info,
+            }
+            .build()
+        };
 
-                        let protected = global.borrow().protected_tags.contains_key(&node.tag);
-                        let transition =
-                            old_state.perform_access(access_kind, rel_pos, protected)?;
-
-                        // Record the event as part of the history
-                        if !transition.is_noop() {
-                            node.debug_info.history.push(diagnostics::Event {
-                                transition,
-                                is_foreign: rel_pos.is_foreign(),
-                                access_cause,
-                                access_range,
-                                transition_range: perms_range.clone(),
-                                span,
-                            });
-                        }
-                        Ok(ContinueTraversal::Recurse)
-                    },
-                    |args: ErrHandlerArgs<'_, TransitionError>| -> InterpError<'tcx> {
-                        let ErrHandlerArgs { error_kind, conflicting_info, accessed_info } = args;
-                        TbError {
-                            conflicting_info,
-                            access_cause,
-                            error_offset: perms_range.start,
-                            error_kind,
-                            accessed_info,
+        if let Some(access_range) = access_range {
+            // Default branch: this is a "normal" access through a known range.
+            // We iterate over affected locations and traverse the tree for each of them.
+            for (perms_range, perms) in self.rperms.iter_mut(access_range.start, access_range.size)
+            {
+                TreeVisitor { nodes: &mut self.nodes, tag_mapping: &self.tag_mapping, perms }
+                    .traverse_parents_this_children_others(
+                        tag,
+                        |args| node_app(perms_range.clone(), args),
+                        |args| err_handler(perms_range.clone(), args),
+                    )?;
+            }
+        } else {
+            // This is a special access through the entire allocation.
+            // It actually only affects `initialized` locations, so we need
+            // to filter on those before initiating the traversal.
+            // In addition this implicit access should not be visible to children,
+            // thus the use of `traverse_nonchildren`.
```

Comment:
This needs an explanation of *why* we skip children. Please mention a testcase that would fail otherwise. (We definitely should have such a testcase.)